internal/api/courses/users/drop.go

```diff
@@ -0,0 +1,55 @@
+package users
+
+import (
+	"github.com/edulinq/autograder/internal/api/core"
+	"github.com/edulinq/autograder/internal/db"
+	"github.com/edulinq/autograder/internal/model"
+)
+
+type DropRequest struct {
+	core.APIRequestCourseUserContext
+	core.MinCourseRoleAdmin
+
+	TargetCourseUser core.TargetCourseUser `json:"target-email"`
+}
+
+type DropResponse struct {
+	FoundUser bool `json:"found-user"`
+}
+
+func HandleDrop(request *DropRequest) (*DropResponse, *core.APIError) {
+	response := DropResponse{}
+
+	if !request.TargetCourseUser.Found {
+		return &response, nil
+	}
+
+	response.FoundUser = true
+
+	if !checkDropPermissions(request) {
+		return nil, core.NewBadCoursePermissionsError("-612", &request.APIRequestCourseUserContext, request.TargetCourseUser.User.Role,
+			"Cannot drop a user with an equal or higher role.").Add("target-course-user", request.TargetCourseUser.User.Email)
```

Comment:
Form the error in the `checkDropPermissions()` function so you can make more specific errors.
Note that the `NewBad*Permissions()` functions hide the message,
so you can be specific without worries.

Reply:
I formed the error in `checkDropPermissions()` but I can't get more specific with the errors. I can't error just because they don't have proper server roles. I have to wait until after the course role check, and then it's a blanket bad perm error because they don't meet either qualification.